File: curbcheck/engine/geo.py
"""The local flat-earth projection every radius query in the engine shares.

Over a 1 km radius at Manhattan's latitude the error against a proper
projection is under a metre, and it keeps pyproj out of the request path.
Values are the WGS-84 metres per degree at 40.75 N.
"""

from __future__ import annotations

import json
import math
from typing import Any

from shapely.errors import ShapelyError
from shapely.geometry import Point, shape

M_PER_DEG_LAT = 111_132.0
M_PER_DEG_LON_AT_EQUATOR = 111_320.0


def meters_per_degree_lon(lat: float) -> float:
    return M_PER_DEG_LON_AT_EQUATOR * math.cos(math.radians(lat))


def degree_padding(lat: float, meters: float) -> tuple[float, float]:
    """The (lon, lat) half-widths of a bbox `meters` wide around a point at `lat`."""
    return (meters / meters_per_degree_lon(lat), meters / M_PER_DEG_LAT)


def to_local_meters(geometry: dict[str, Any], *, lon0: float, lat0: float) -> dict[str, Any]:
    """The same GeoJSON geometry with (lon, lat) replaced by metres from (lon0, lat0)."""
    scale_lon = meters_per_degree_lon(lat0)

    def convert(node: Any) -> Any:
        if isinstance(node, (list, tuple)) and node and isinstance(node[0], (int, float)):
            return [(float(node[0]) - lon0) * scale_lon, (float(node[1]) - lat0) * M_PER_DEG_LAT]
        return [convert(child) for child in node]

    return {"type": geometry["type"], "coordinates": convert(geometry["coordinates"])}


def measure(
    geom: Any, *, lon: float, lat: float, origin: Point
) -> tuple[dict[str, Any], float] | None:
    """A geometry column's value and its distance in metres, or None when it cannot be read.

    The database is untrusted at read time (CLAUDE.md): a truncated or
    hand-edited snapshot answered every search with a 500 before the guard
    (docs/SECURITY.md residual 9). `origin` is Point(0, 0), the query point in
    the local frame; callers hoist it out of their loop.
    """
    try:
        geometry = json.loads(str(geom))
        local = shape(to_local_meters(geometry, lon0=lon, lat0=lat))
        return geometry, float(local.distance(origin))
    except (ValueError, TypeError, KeyError, IndexError, ShapelyError):
        return None

File: site/static/engine/geo.js
/**
 * The local flat-earth projection every radius query in the engine shares.
 *
 * Port of `curbcheck/engine/geo.py`. Over a 1 km radius at Manhattan's latitude
 * the error against a proper projection is under a metre. Values are the WGS-84
 * metres per degree at 40.75 N.
 *
 * The one difference from the Python: a geometry arrives as a row of a packed
 * coords/offsets column rather than as a GeoJSON string, so `measure` returns
 * only the distance — the caller that needs the geometry asks `pack.lineString`
 * for it.
 */

export const M_PER_DEG_LAT = 111_132.0;
export const M_PER_DEG_LON_AT_EQUATOR = 111_320.0;

// `math.radians(deg)` is `deg * (pi / 180)`; the parentheses keep the constant
// the one Python multiplies by, so the product is bit-for-bit the same double.
const RADIANS_PER_DEGREE = Math.PI / 180;

/**
 * @param {number} lat
 * @returns {number}
 */
export function metersPerDegreeLon(lat) {
  return M_PER_DEG_LON_AT_EQUATOR * Math.cos(lat * RADIANS_PER_DEGREE);
}

/**
 * The (lon, lat) half-widths of a bbox `meters` wide around a point at `lat`.
 *
 * @param {number} lat
 * @param {number} meters
 * @returns {{lon: number, lat: number}}
 */
export function degreePadding(lat, meters) {
  return { lon: meters / metersPerDegreeLon(lat), lat: meters / M_PER_DEG_LAT };
}

/**
 * One row of a coords column with (lon, lat) replaced by metres from (lon0, lat0).
 *
 * Null when the row holds fewer than two vertices, which is what shapely
 * refuses to build a LineString from — `data/` is untrusted at read time
 * (CLAUDE.md) and the pack is read the same way the database was.
 *
 * @param {{coords: ArrayLike<number>, offsets: ArrayLike<number>}} geom
 * @param {number} row
 * @param {number} lon0
 * @param {number} lat0
 * @returns {Float64Array | null}
 */
export function toLocalMeters(geom, row, lon0, lat0) {
  if (row < 0 || row + 1 >= geom.offsets.length) {
    return null;
  }
  const start = geom.offsets[row];
  const end = geom.offsets[row + 1];
  if (end - start < 2) {
    return null;
  }
  const scaleLon = metersPerDegreeLon(lat0);
  const local = new Float64Array((end - start) * 2);
  for (let vertex = start; vertex < end; vertex += 1) {
    const at = (vertex - start) * 2;
    local[at] = (geom.coords[2 * vertex] - lon0) * scaleLon;
    local[at + 1] = (geom.coords[2 * vertex + 1] - lat0) * M_PER_DEG_LAT;
  }
  return local;
}

/**
 * The distance in metres from (lon, lat) to one geometry row, or null when it
 * cannot be read.
 *
 * The query point is the origin of the local frame, which is what the Python
 * passes as `Point(0, 0)`.
 *
 * @param {{coords: ArrayLike<number>, offsets: ArrayLike<number>}} geom
 * @param {number} row
 * @param {number} lon
 * @param {number} lat
 * @returns {{distanceM: number} | null}
 */
export function measure(geom, row, lon, lat) {
  const local = toLocalMeters(geom, row, lon, lat);
  if (local === null) {
    return null;
  }
  let nearest = Infinity;
  for (let at = 0; at + 3 < local.length; at += 2) {
    const distance = pointToSegment(0, 0, local[at], local[at + 1], local[at + 2], local[at + 3]);
    if (distance < nearest) {
      nearest = distance;
    }
  }
  if (!Number.isFinite(nearest)) {
    return null;
  }
  return { distanceM: nearest };
}

/**
 * GEOS's `Distance::pointToSegment`, written out so the two engines agree to
 * the ulp: shapely's `distance` is this function minimised over the segments of
 * the line, and the differential harness compares distances exactly.
 *
 * @param {number} px
 * @param {number} py
 * @param {number} ax
 * @param {number} ay
 * @param {number} bx
 * @param {number} by
 * @returns {number}
 */
export function pointToSegment(px, py, ax, ay, bx, by) {
  // GEOS: if start = end, the segment is a point, so use one endpoint.
  if (ax === bx && ay === by) {
    return distanceToPoint(px, py, ax, ay);
  }
  const dx = bx - ax;
  const dy = by - ay;
  const len2 = dx * dx + dy * dy;
  const r = ((px - ax) * dx + (py - ay) * dy) / len2;
  if (r <= 0.0) {
    return distanceToPoint(px, py, ax, ay);
  }
  if (r >= 1.0) {
    return distanceToPoint(px, py, bx, by);
  }
  const s = ((ay - py) * dx - (ax - px) * dy) / len2;
  return Math.abs(s) * Math.sqrt(len2);
}

/** `Coordinate::distance`: sqrt of the sum of the squared deltas, not `Math.hypot`. */
function distanceToPoint(px, py, ax, ay) {
  const ddx = px - ax;
  const ddy = py - ay;
  return Math.sqrt(ddx * ddx + ddy * ddy);
}
